libtransmission/rpc-server.cc

```diff
@@ -77,39 +77,101 @@ auto inline constexpr TrUnixAddrStrLen = size_t{ sizeof(((struct sockaddr_un*)nu
 
 enum tr_rpc_address_type
 {
-    TR_RPC_AF_INET,
-    TR_RPC_AF_INET6,
-    TR_RPC_AF_UNIX
+    TR_RPC_UNIX_ADDR,
+    TR_RPC_INET_ADDR,
+    TR_RPC_NUM_ADDR
+};
+
+struct tr_unix_addr
+{
+    std::array<char, TrUnixAddrStrLen> unixSocketPath;
+
+    std::string to_string()
+    {
+        if (auto str = unixSocketPath.data())
+        {
+            return { str };
+        }
+        return std::string{ TrUnixSocketPrefix };
+    }
+
+    bool from_string(std::string_view src)
+    {
+        if (std::size(src) >= TrUnixAddrStrLen)
+        {
+            tr_logAddError(fmt::format(
+                _("Unix socket path must be fewer than {count} characters (including '{prefix}' prefix)"),
+                fmt::arg("count", TrUnixAddrStrLen - 1),
+                fmt::arg("prefix", TrUnixSocketPrefix)));
+            return false;
+        }
+        tr_strlcpy(std::data(unixSocketPath), std::string{ src }.c_str(), std::size(unixSocketPath));
+        return true;
+    }
```

Comment:
- this should be a class rather than a struct, with private field & public method

- given the extra complexity of strlcpy below and this data check here, maybe it would make more sense to have `unixSocketPath` be a `std::string`?

- All of libtransmission is slowly moving to `snake_case` naming for methods and `snake_case_` for class fields, so let's do that here with `unix_socket_path_`

- It's a little odd that half of the sanity checks are below (does it begin with the right prefix?) below and half are here (is the path too long?). Let's consolidate them here
 
```suggestion
class tr_unix_addr
{
public:
    [[nodiscard]] std::string to_string() const
    {
		    return std::empty(unix_socket_path_) ? unix_socket_path_ : TrUnixSocketPrefix;
    }

    [[nodiscard]] bool from_string(std::string_view src)
    {
        if (tr_strvStartsWith(TrUnixSocketPrefix, src))
        {
				    return false;
        }

        if (std::size(src) >= TrUnixAddrStrLen)
        {
            tr_logAddError(fmt::format(
                _("Unix socket path must be fewer than {count} characters (including '{prefix}' prefix)"),
                fmt::arg("count", TrUnixAddrStrLen - 1),
                fmt::arg("prefix", TrUnixSocketPrefix)));
            return false;
        }

				unix_socket_path_ = src;
        return true;
    }

private:
    std::string unix_socket_path_;
```

Reply:
Most of that makes sense, I guess, but I'm still a little fuzzy on struct vs class. Should `tr_rpc_address` be a class as well? Again, I tried to mirror `tr_address` here, which is declared as a struct.

Reply:
@ckerr I don't think you can mix non-trivial constructors in anonymous unions, hence struct is needed here. I'm not entirely sure though.

Reply:
TBH at this point `tr_address` could probably be a class too. It's been a struct for so long because it made it easier to slowly migrate the codebase from C to C++, but at this point the difference is moot.

Yes I'm pretty sure you're right about nontrivial constructors, I forgot about that. We could use a std::array again or get _really_ C++ish and replace the union with a std::variant. Either approach wfm